Labrad data support

xlelephant opened the pull request:
This PR aims to address https://github.com/pyplotter/pyplotter/issues/7 by porting the Labrad backend from the repository provided [here](https://github.com/labrad/servers/tree/master/datavault).
Each folder ending with ".dir" is treated as a Labrad "Database". Additionally, I add double-click to navigation to the parent folder if datasets are stored within a child folder. To efficiently handle large datasets, I utilize dv.listDatasets() to check database items quickly without loading dataset information.

loadLabradDataFromRun.py is responsible for parsing parameters and loading data from a dataset. Notably, to expedite the plotting process, I've opted not to employ mp.Queue here. The performance is almost comparable to Labrad's Grapher.exe.)

Default Labrad backend settings have been adjusted to open HDF5 files with the Single-Writer/Multiple-Reader (SWMR) feature enabled for improved efficiency.

xlelephant committed:
Labrad data support, first try

edumur commented:
Hi,

I currently do not have time to check such amount of new code.
I will be able to check this in ~2 weeks, I think.

Thanks for the work!

xlelephant commented:
> Hi,
> 
> I currently do not have time to check such amount of new code. I will be able to check this in ~2 weeks, I think.
> 
> Thanks for the work!

No problem, take your time.
Thank you!

xlelephant committed:
Close old livePlots.

LivePlot exceeds the maximum number, which is set in config["livePlotWindowNumber"], will be closed automatically.

xlelephant committed:
Auto close 2D live plots when > livePlotWindowNumber

xlelephant committed:
Plot improvements

add "plotHideInteractionPanel" to auto hide the interaction panel
Fix IQ data, with no independent
Add close all plots

xlelephant committed:
Change ui layout
improve layout
Fix bugs in manual plot

edumur committed:
Solve #3 by replacing `cumtrapz` by `cumulative_trapezoid`

edumur commented on pyplotter/pyplotter.py:
```diff
@@ -6,6 +6,7 @@
 # Init config file, to be done first
 initConfig()
 config = loadConfigCurrent()
+os.environ['HDF5_USE_FILE_LOCKING'] = config['HDF5_USE_FILE_LOCKING']
```
Please add a comment here (basically the same one as in the config file will suffice)

xlelephant replied:
I removed this line.

edumur commented on pyplotter/sources/functions.py:
```diff
@@ -118,6 +113,16 @@ def isBlueForsFolder(folderName : str) -> bool:
     return len(folderName.split('-'))==3 and all([len(i)==2 for i in folderName.split('-')])
 
 
+def isQcodesData(folderName: str):
```
`isQcodesData(folderName: str) -> bool:`

And add docstring ;) .

edumur replied:
Still waiting the return type and the docstring.

xlelephant replied:
added.

edumur commented on pyplotter/sources/functions.py:
```diff
@@ -203,6 +205,8 @@ def getPlotRef(databaseAbsPath: str,
     # If BlueFors log files
     if isBlueForsFolder(currentPath):
         dataPath = currentPath
+    elif isLabradFolder(databaseAbsPath):
```
add `# If Labrad data folder`

edumur replied:
Still there

xlelephant replied:
added.

edumur commented on pyplotter/sources/functions.py:
```diff
@@ -606,3 +601,78 @@ def getDialogWidthHeight(nbDialog: int) -> Tuple[List[int]]:
         ys[i] = int(i//2*dialogHeight)
 
     return xs, ys, dialogWidths, dialogHeights
+
```
This  looks like is not related to labrad but to how livePlot dialog windows are handles.
Could you explain why you changed the code here?

edumur replied:
Still waiting an explanation ^^

xlelephant replied:
Explained in the docstring.

edumur commented on pyplotter/sources/functions.py:
```diff
@@ -606,3 +601,78 @@ def getDialogWidthHeight(nbDialog: int) -> Tuple[List[int]]:
+    return xs, ys, dialogWidths, dialogHeights
+
+
+def plotIdGenerator():
```
What is the purpose of this function?

xlelephant replied:
also explained.

edumur commented on pyplotter/sources/labradDatavault.py:
```diff
@@ -0,0 +1,259 @@
+import multiprocess as mp
+import copy
+import labrad
+from typing import Tuple, List  # , Dict,  Union, Optional, final
```
Please remove unused comment

xlelephant replied:
removed.

edumur commented on pyplotter/sources/labradDatavault.py:
```diff
@@ -0,0 +1,259 @@
+# you might want to specify the correct labrad host and password
+
+
+def name2idx(variables, name, default=np.nan):
```
If you could add docstring and typing to your functions?

xlelephant replied:
Added necessary docstring and typing.

edumur commented on pyplotter/sources/labradDatavault.py:
```diff
@@ -0,0 +1,259 @@
+        return dep["name"]
+
+
+def getNbTotalRun(databaseAbsPath: str) -> None:
```
return int no?

xlelephant replied:
changed to int.

edumur commented on pyplotter/sources/labradDatavault.py:
```diff
@@ -0,0 +1,259 @@
+    return nbTotalRun
+
+
+def getNbTotalRunmp(databaseAbsPath: str, queueNbRun: mp.Queue):
```
return None

xlelephant replied:
added return None.

edumur commented on pyplotter/sources/workers/loadLabradDataBase.py:
```diff
@@ -0,0 +1,174 @@
+from PyQt5 import QtCore
+import multiprocess as mp
```
not used

xlelephant replied:
removed this file.

edumur commented on pyplotter/ui/dialogs/dialogLiveplot.py:
```diff
@@ -1,26 +1,41 @@
 from __future__ import annotations
 from PyQt5 import QtWidgets, QtCore
+from idna import check_bidi
```
not used, please remove

xlelephant replied:
removed.

edumur commented on pyplotter/ui/tableWidgetDatabase.py:
```diff
@@ -8,18 +8,56 @@
 PICTURESPATH = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'pictures')
 
 
+def LoadDataBaseThread(databaseAbsPath, progressBar):
```
I would prefer not to override the function here but to use the condition later on in the code.
Also `isMongoDBFolder` is undefined.

xlelephant replied:
Good! I improved these according to your suggestion.

edumur commented:
Hi,

First I apology for the delay.
Second, thanks for the big work you did.

As I can't check labrad features locally I will trust you on that.

I have few comments that I gave you directly on code.
Can you please check them?

General comments:
1. Please add labrad in the setup file and in the lib description 
2. Since you're faking the progressBar in the `loadLabradDatabase` file, could you just load everything at once without thread like for csv files?
3. I can't run labrad on my computer even though I will not use it:
 ```bash
plotter
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\ed261978\Anaconda3\envs\pyplotter2\Scripts\plotter.exe\__main__.py", line 4, in <module>
  File "C:\Users\ed261978\Desktop\edumur\scripts\pyplotter2\pyplotter\pyplotter.py", line 10, in <module>
    from .ui.mainWindow import MainApp
  File "C:\Users\ed261978\Desktop\edumur\scripts\pyplotter2\pyplotter\ui\mainWindow.py", line 7, in <module>
    from .mainWindowui_r2c2 import Ui_MainWindow
  File "C:\Users\ed261978\Desktop\edumur\scripts\pyplotter2\pyplotter\ui\mainWindowui_r2c2.py", line 352, in <module>
    from ..ui.menuBar import MenuBar
  File "C:\Users\ed261978\Desktop\edumur\scripts\pyplotter2\pyplotter\ui\menuBar.py", line 11, in <module>
    from .dialogs.dialogLiveplot import DialogLiveplot
  File "C:\Users\ed261978\Desktop\edumur\scripts\pyplotter2\pyplotter\ui\dialogs\dialogLiveplot.py", line 12, in <module>
    from ...sources.labradDatavault import (
  File "C:\Users\ed261978\Desktop\edumur\scripts\pyplotter2\pyplotter\sources\labradDatavault.py", line 8, in <module>
    CXN = labrad.connect()
          ^^^^^^^^^^^^^^^^
  File "C:\Users\ed261978\Anaconda3\envs\pyplotter2\Lib\site-packages\labrad\__init__.py", line 42, in connect
    cxn = backend.connect(host=host, port=port, name=name, **kw)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ed261978\Anaconda3\envs\pyplotter2\Lib\site-packages\labrad\backend.py", line 86, in connect
    cxn = future.result(timeout=timeout)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ed261978\Anaconda3\envs\pyplotter2\Lib\concurrent\futures\_base.py", line 456, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\ed261978\Anaconda3\envs\pyplotter2\Lib\concurrent\futures\_base.py", line 401, in __get_result
    raise self._exception
  File "C:\Users\ed261978\Anaconda3\envs\pyplotter2\Lib\site-packages\labrad\concurrent.py", line 84, in wrapped
    result = yield defer.maybeDeferred(func, *args, **kw)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
twisted.internet.error.ConnectionRefusedError: Connection was refused by other side: 10061: No connection could be made because the target machine actively refused it..
```
This is a major issue as it should work for non labrad users.
5. I was then unable to see how and why you change the liveplot dialog and the mainWindow ui.


Thanks!

edumur committed:
Update widgetCSV.py

edumur committed:
Solve latex equation display

edumur committed:
fix BlueFors log issue #11

edumur committed:
Improve livePlot for 2d, see #13

edumur committed:
Fix error on MacOs, see #13

xlelephant committed:
Fix import errors
Add package requirements
support local Labrad dataVault

edumur committed:
LivePlot works with donD qcode functions

edumur committed:
remove tentative fix for macOS

edumur committed:
upgrade version

xlelephant committed:
load LabradDatabase at once without thread

xlelephant committed:
add Labrad in the lib des, readme, requirements

xlelephant committed:
revert to default layout; small fixes

xlelephant commented:
Thanks for reviewing! 

The initial goal was to open Labrad data without connecting the server.
However, I was a bit trapped in testing "SWMR" mode of hdf5 files, as it turned out not as stable as we expected.
Instead, I provide two ways in this PR to read Labrad datasets. 
One is "ONLY" for offline and local file reading; 
Another is to use the Labrad server, like a normal Client, which guarantees no harm to the live data.
See `Readme.md`

> 1. Please add labrad in the setup file and in the lib description

Re: I added pylabrad descriptions and also added a requirements.txt file.

> 2. Since you're faking the progressBar in the loadLabradDatabase file, could you just load everything at once without thread like for csv files?

Re: Good suggestion! `loadLabradDatabase` is removed now. The signal flow is much simpler now: 
![image](https://github.com/pyplotter/pyplotter/assets/50734064/27d90d94-67e9-47c1-ae06-7a928896272f)

> 3. I can't run labrad on my computer even though I will not use it:
> This is a major issue as it should work for non labrad users.

Re: Non-Labrad users should not be affected now, even if they choose not to install the pylabrad package.

> 4. I was then unable to see how and why you change the liveplot dialog and the mainWindow ui.

Re: I revert to the default layout. You can also have a look at the "Labrad-type" layout by using `mainWindowui_r2c2.py`;)
This layout may not look well on high-dpi screens, I could do more tweaks.

xlelephant committed:
fix high-dpi display

by changing maximum width and height of text

xlelephant committed:
clean up

edumur commented on README.md:
```diff
@@ -22,6 +22,7 @@ Currently the following packages are required:
 * qcodes>=0.26.0
 * scikit-rf
 * scipy
+* pylabrad (Labrad data support)
```
I am currently  in favor of putting pylabrad as an optional requirement.
Please update the readme accordingly.

Also the setup.cfg would have to be modified following: [](https://setuptools.pypa.io/en/latest/userguide/declarative_config.html)
Adding an optional section
```
[options.extras_require]
```

xlelephant replied:
I added extras_require filed, but not fully tested the functionality yet.

edumur commented on pyplotter/sources/functions.py:
```diff
@@ -247,10 +250,11 @@ def getPlotTitle(databaseAbsPath: str,
         return os.path.basename(databaseAbsPath)[:-3]
 
+def getLabradDataVaultNameFromAbsPath(databaseAbsPath: str) -> str:
+        return os.path.basename(databaseAbsPath)[:-4]
```
Both `getDatabaseNameFromAbsPath` and `getLabradDataVaultNameFromAbsPath` have their return too much indented

xlelephant replied:
I merged both of them in one function, which is indented correctly.

edumur commented on pyplotter/sources/labradDatavault.py:
```diff
@@ -0,0 +1,479 @@
+import glob
```
not used

xlelephant replied:
removed.

edumur commented on pyplotter/ui/tableWidgetFolder.py:
```diff
@@ -203,22 +202,36 @@ def itemClicked_(self, b: Union[int, QtCore.QPoint]) -> None:
             # If it is a npz file
             elif nextPath[-3:].lower() in 'npz':
                 self.npzClick()
+            elif isinstance(b, QtCore.QPoint):
```
Could you explain in which situation you need this?

xlelephant replied:
This is redundant, removed.

edumur commented on pyplotter/ui/tableWidgetParameter.py:
```diff
@@ -1,15 +1,77 @@
+        labrad_dataset = LabradDataset(databaseAbsPath, noisy=False)
+        labrad_dataset.loadDataset(runId)
+        return labrad_dataset 
+    # elif isMongoDBFolder(databaseAbsPath):
```
Please remove

xlelephant replied:
removed.

edumur reviewed (commented):
Hi,

Thanks for your work.
The plotter is running locally even though I still didn't run extensive tests.
Please solve the current pending reviews.

Global remarks:

At least locally I have the following error when double clicking on a run:
```bash
Traceback (most recent call last):
  File "C:\Users\ed261978\Desktop\edumur\scripts\pyplotter\pyplotter\ui\tableWidgetParameter.py", line 603, in slotFillTableWidgetParameter
    plotTitle   = getPlotTitle(databaseAbsPath, runId, self.dataset.name)
                                                       ^^^^^^^^^^^^^^^^^
AttributeError: 'function' object has no attribute 'name'
```
Could be related to the change you made in the `tableWidgetParameter` file line 601.

Your two functions are not used in the same way, could you fixe it:
```
isLabradFolder(databaseAbsPath)
isQcodesData(os.path.split(databaseAbsPath)[-1])
```

xlelephant committed:
solve double clicking on qcode data

clean up tableWidgetParameter.py

xlelephant committed:
add docstrings in function.py

xlelephant committed:
modify Labrad as Optional requirement.

xlelephant committed:
add docstrings for labradDatavault.py

xlelephant committed:
small fixes

xlelephant committed:
Fix Labrad data live plot, two modes tested

xlelephant committed:
live plot window number default to 2,

and enable plotHideInteractionPanel

xlelephant committed:
Merge branch 'pyplotter:master' into feature_LabradData

xlelephant committed:
use database pattern functions in the same way

xlelephant committed:
fix tableWidgetDatabase.eventFilter warning

xlelephant committed:
add more docstring

xlelephant commented:
Thanks for the code review! I refactored the files that you redlighted.
Two Labrad data reading methods were also tested.

xlelephant committed:
remove a file.

xlelephant committed:
fix LocalDatavault.ls() ordering problem for different filesystem.

edumur commented on pyplotter/sources/labradDatavault.py:
```diff
@@ -0,0 +1,525 @@
+MAX_NUM_PLOT_DEPS = 8
+
+
+def labrad_urldecode(data_url) -> any:
```
The correct typing is `Any` starting with an uppercase.

edumur commented on setup.cfg:
```diff
@@ -38,6 +38,10 @@ install_requires =
 
 
+[options.extras_require]
+larbad = pylabrad
```
Typo
`labrad = pylabrad`

edumur reviewed (commented):
Thanks again for all your work.

Some remarks:
1. the doubleClick on a run of qcodes database is not working. I just disable the widget row and no plot is displayed,
2. By default the plots hide the interaction panel, please leave it on by default. You can add an option in the menu bar, maybe in preference? About it if you wish.


Would you accept to be "in charge" of labrad thing? I have no idea how labrad works...

xlelephant committed:
revert to opening the interaction panel, fix typo

xlelephant commented:
> Thanks again for all your work.
> 
> Some remarks:
> 
> 1. the doubleClick on a run of qcodes database is not working. I just disable the widget row and no plot is displayed,

DoubleClick on a Labrad dataset for quick plotting is wanted by most Labrad users, as they can easily check the data. 
But I did not change the default behavior of that on qcodes database, do you want me also to add this feature to qcodes, OR add a statement no the Tip section of the README?

> 2. By default the plots hide the interaction panel, please leave it on by default. You can add an option in the menu bar, maybe in preference? About it if you wish.

I revert that to show the interaction panel by default. Adding a dynamic option in the menu bar is sweet, but I opt not to add too many buttons on the GUI. Instead, I can add explanations for various settings in README.

> 
> Would you accept to be "in charge" of labrad thing? I have no idea how labrad works...

Sure, I would like to do that, and perhaps future development on general database format based on HDF5/NetCDF files.

edumur commented:
Arf sorry it was a typo... What I meant was that the doubleClick on a run of qcodes database is not working. When I adoubleClik on a run, it just disable the widget row and no plot is displayed,

xlelephant committed:
Fix doubleClick on a run of qcodes database.

Add 'progressBarMaxNb' to avoid displaying too many progressBar.

xlelephant committed:
doubleClick fix

xlelephant commented:
> Arf sorry it was a typo... What I meant was that the doubleClick on a run of qcodes database is not working. When I adoubleClik on a run, it just disable the widget row and no plot is displayed,

Oh sorry, I deleted this by mistake. 
Fixed now, please check, thanks!

edumur commented:
Hi,

Yep, doubleClick is now working.

I saw you added a config parameter `progressBarMaxNb`.
What happen when an user launched a 4th progressBar which then is not created?
How does the program handles the update and deletion of the non existing progressBar?

edumur commented:
Can you add yourself on the authors section  of the readme file as the Labrad guy?
Thanks!

xlelephant committed:
add author and acknowledgment in readme

xlelephant commented:
> Hi,
> 
> Yep, doubleClick is now working.
> 
> I saw you added a config parameter `progressBarMaxNb`. What happen when an user launched a 4th progressBar which then is not created? How does the program handles the update and deletion of the non existing progressBar?

If too many dependents are toggled at once(to plot), the layout will get too wide to fill in all these progressBars, so I added this restriction.
Except for being added as the "addPermanentWidget", the 4th progressBar will be updated and poped normally.
Note that the `removeWidget` function completes silently even if the input widget was not added.

xlelephant commented:
> Can you add yourself on the authors section of the readme file as the Labrad guy? Thanks!

Added, thanks!
